File: design-pattern/Creational Design Patterns/factory-design/SimpleFactory.py
from abc import ABC, abstractmethod

class Burger(ABC):
    @abstractmethod
    def prepare(self):
        pass


class BasicBurger(Burger):
    def prepare(self):
        print("Basic burger is prepared")

class StandardBurger(Burger):
    def prepare(self):
        print("Standard burger is prepared")

class PreminumBurger(Burger):
    def prepare(self):
        print("Preminum burger is created")

class BurgerFactory():
    def createBurger(self, type:str)-> Burger:
        if type.lower() == 'basic':
            return BasicBurger()
        elif type.lower() == 'standard':
            return StandardBurger()
        elif type.lower() == 'premium':
            return PreminumBurger()
        else:
            print('Invalid Burger')

def main():
    type = "Standard"

    obj = BurgerFactory()
    burger = obj.createBurger(type)
    if burger:
        burger.prepare();

if __name__ == "main":
    main()

File: design-pattern/Creational Design Patterns/factory-design/SimpleFactory.cpp
#include <iostream>
#include <string>

class Burger
{
public:
    virtual void prepare() = 0;
    virtual ~Burger() {}
};

class BasicBurger : public Burger
{
public:
    void prepare() override
    {
        std::cout << "Basic burger prepared";
    }
};

class StandardBurger : public Burger
{
public:
    void prepare() override
    {
        std::cout << "Standard Burger Prepared";
    }
};

class PremiumBurger : public Burger
{
public:
    void prepare() override
    {
        std::cout << "Premium burger prepared";
    }
};
class BurgerFactory
{
public:
    Burger *createBurger(std::string &type)
    {
        if (type == "basic")
        {
            return new BasicBurger();
        }
        else if (type == "standard")
        {
            return new StandardBurger();
        }
        else if (type == "premium")
        {
            return new PremiumBurger();
        }
        else
        {
            std::cout << "Invalid Burger Type" << "\n";
            return nullptr;
        }
    }
};

int main()
{
    std::string type = "standard";

    BurgerFactory *orderBurger = new BurgerFactory();
    Burger *burger = orderBurger->createBurger(type);

    burger->prepare();

    return 0;
}